{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "0cbd2272",
   "metadata": {},
   "source": [
    "# Machine Learning for Cancer Prediction"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8ce14227",
   "metadata": {},
   "source": [
    "This collection of data is part of the RNA-Seq (HiSeq) PANCAN data set, it is a random extraction of gene expressions of patients having different types of tumor:\n",
    "\n",
    "* BRCA: Breast invasive carcinoma\n",
    "        BRCA is a type of cancer that starts in the cells of the breast. It is one of the most common cancers among women worldwide and can occur in both men and women, although it is much more common in women.\n",
    "* KIRC: Kidney renal clear cell carcinoma\n",
    "        KIRC is a type of kidney cancer that begins in the lining of the kidney tubules. It is the most common type of kidney cancer, accounting for about 70-75% of all kidney cancer cases.\n",
    "* COAD: Colon adenocarcinoma\n",
    "        COAD is a type of cancer that develops in the colon or rectum and arises from the glands in the lining of the colon and rectum. It is one of the most common forms of colorectal cancer.\n",
    "* LUAD: Lung adenocarcinoma\n",
    "        LUAD is a type of non-small cell lung cancer that begins in the cells that line the airways of the lungs. It is the most common type of lung cancer, accounting for about 40% of lung cancer cases.\n",
    "* PRAD: Prostate adenocarcinoma\n",
    "        PRAD is a type of cancer that develops in the prostate gland, which is a part of the male reproductive system. It is one of the most common cancers among men and usually grows slowly, but it can spread to other parts of the body if not treated early."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "01cd7bc3",
   "metadata": {},
   "source": [
    "Variable Information\n",
    "A dummy name (gene_XX) is given to each attribute. Check the original submission (https://www.synapse.org/#!Synapse:syn4301332), or the platform specs for the complete list of probes name. The attributes are ordered consitently with the original submission.\n",
    "\n",
    "Samples (instances) are stored row-wise. Variables (attributes) of each sample are RNA-Seq gene expression levels measured by illumina HiSeq platform."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "23bbff1e",
   "metadata": {},
   "source": [
    "# Import Python Libraries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "6fa6ba1f",
   "metadata": {},
   "outputs": [],
   "source": [
    "#data handling\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "\n",
    "#data visualization\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "\n",
    "#preprocessing\n",
    "from sklearn.model_selection import train_test_split\n",
    "from sklearn.preprocessing import LabelEncoder\n",
    "from sklearn.preprocessing import label_binarize\n",
    "from sklearn.preprocessing import MinMaxScaler\n",
    "\n",
    "#feature selection\n",
    "from sklearn.feature_selection import mutual_info_classif\n",
    "\n",
    "#classification\n",
    "from sklearn.multiclass import OneVsRestClassifier\n",
    "from sklearn.ensemble import RandomForestClassifier\n",
    "\n",
    "# performance metrics\n",
    "from sklearn.metrics import balanced_accuracy_score,f1_score,precision_score, recall_score\n",
    "from sklearn.metrics import classification_report\n",
    "from sklearn.metrics import confusion_matrix\n",
    "#from sklearn.metrics import plot_confusion_matrix\n",
    "from sklearn.metrics import roc_curve,auc\n",
    "from sklearn.metrics import roc_auc_score\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "90b1074d",
   "metadata": {},
   "source": [
    "# Read in the Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "b87636c4",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.read_csv('C:/Users/arutherford/Documents/Python Scripts/Dataset/UCI_GeneExpression.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "7da801f5",
   "metadata": {},
   "outputs": [],
   "source": [
    "labels = pd.read_csv('C:/Users/arutherford/Documents/Python Scripts/Dataset/UCI_GeneExpression_labels.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "f1ca0de8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>gene_0</th>\n",
       "      <th>gene_1</th>\n",
       "      <th>gene_2</th>\n",
       "      <th>gene_3</th>\n",
       "      <th>gene_4</th>\n",
       "      <th>gene_5</th>\n",
       "      <th>gene_6</th>\n",
       "      <th>gene_7</th>\n",
       "      <th>gene_8</th>\n",
       "      <th>...</th>\n",
       "      <th>gene_20521</th>\n",
       "      <th>gene_20522</th>\n",
       "      <th>gene_20523</th>\n",
       "      <th>gene_20524</th>\n",
       "      <th>gene_20525</th>\n",
       "      <th>gene_20526</th>\n",
       "      <th>gene_20527</th>\n",
       "      <th>gene_20528</th>\n",
       "      <th>gene_20529</th>\n",
       "      <th>gene_20530</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>sample_0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.017209</td>\n",
       "      <td>3.265527</td>\n",
       "      <td>5.478487</td>\n",
       "      <td>10.431999</td>\n",
       "      <td>0.0</td>\n",
       "      <td>7.175175</td>\n",
       "      <td>0.591871</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>4.926711</td>\n",
       "      <td>8.210257</td>\n",
       "      <td>9.723516</td>\n",
       "      <td>7.220030</td>\n",
       "      <td>9.119813</td>\n",
       "      <td>12.003135</td>\n",
       "      <td>9.650743</td>\n",
       "      <td>8.921326</td>\n",
       "      <td>5.286759</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>sample_1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.592732</td>\n",
       "      <td>1.588421</td>\n",
       "      <td>7.586157</td>\n",
       "      <td>9.623011</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.816049</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>4.593372</td>\n",
       "      <td>7.323865</td>\n",
       "      <td>9.740931</td>\n",
       "      <td>6.256586</td>\n",
       "      <td>8.381612</td>\n",
       "      <td>12.674552</td>\n",
       "      <td>10.517059</td>\n",
       "      <td>9.397854</td>\n",
       "      <td>2.094168</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>sample_2</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.511759</td>\n",
       "      <td>4.327199</td>\n",
       "      <td>6.881787</td>\n",
       "      <td>9.870730</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.972130</td>\n",
       "      <td>0.452595</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>5.125213</td>\n",
       "      <td>8.127123</td>\n",
       "      <td>10.908640</td>\n",
       "      <td>5.401607</td>\n",
       "      <td>9.911597</td>\n",
       "      <td>9.045255</td>\n",
       "      <td>9.788359</td>\n",
       "      <td>10.090470</td>\n",
       "      <td>1.683023</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>sample_3</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.663618</td>\n",
       "      <td>4.507649</td>\n",
       "      <td>6.659068</td>\n",
       "      <td>10.196184</td>\n",
       "      <td>0.0</td>\n",
       "      <td>7.843375</td>\n",
       "      <td>0.434882</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>6.076566</td>\n",
       "      <td>8.792959</td>\n",
       "      <td>10.141520</td>\n",
       "      <td>8.942805</td>\n",
       "      <td>9.601208</td>\n",
       "      <td>11.392682</td>\n",
       "      <td>9.694814</td>\n",
       "      <td>9.684365</td>\n",
       "      <td>3.292001</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>sample_4</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.655741</td>\n",
       "      <td>2.821547</td>\n",
       "      <td>6.539454</td>\n",
       "      <td>9.738265</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.566967</td>\n",
       "      <td>0.360982</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>5.996032</td>\n",
       "      <td>8.891425</td>\n",
       "      <td>10.373790</td>\n",
       "      <td>7.181162</td>\n",
       "      <td>9.846910</td>\n",
       "      <td>11.922439</td>\n",
       "      <td>9.217749</td>\n",
       "      <td>9.461191</td>\n",
       "      <td>5.110372</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 20532 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  Unnamed: 0  gene_0    gene_1    gene_2    gene_3     gene_4  gene_5  \\\n",
       "0   sample_0     0.0  2.017209  3.265527  5.478487  10.431999     0.0   \n",
       "1   sample_1     0.0  0.592732  1.588421  7.586157   9.623011     0.0   \n",
       "2   sample_2     0.0  3.511759  4.327199  6.881787   9.870730     0.0   \n",
       "3   sample_3     0.0  3.663618  4.507649  6.659068  10.196184     0.0   \n",
       "4   sample_4     0.0  2.655741  2.821547  6.539454   9.738265     0.0   \n",
       "\n",
       "     gene_6    gene_7  gene_8  ...  gene_20521  gene_20522  gene_20523  \\\n",
       "0  7.175175  0.591871     0.0  ...    4.926711    8.210257    9.723516   \n",
       "1  6.816049  0.000000     0.0  ...    4.593372    7.323865    9.740931   \n",
       "2  6.972130  0.452595     0.0  ...    5.125213    8.127123   10.908640   \n",
       "3  7.843375  0.434882     0.0  ...    6.076566    8.792959   10.141520   \n",
       "4  6.566967  0.360982     0.0  ...    5.996032    8.891425   10.373790   \n",
       "\n",
       "   gene_20524  gene_20525  gene_20526  gene_20527  gene_20528  gene_20529  \\\n",
       "0    7.220030    9.119813   12.003135    9.650743    8.921326    5.286759   \n",
       "1    6.256586    8.381612   12.674552   10.517059    9.397854    2.094168   \n",
       "2    5.401607    9.911597    9.045255    9.788359   10.090470    1.683023   \n",
       "3    8.942805    9.601208   11.392682    9.694814    9.684365    3.292001   \n",
       "4    7.181162    9.846910   11.922439    9.217749    9.461191    5.110372   \n",
       "\n",
       "   gene_20530  \n",
       "0         0.0  \n",
       "1         0.0  \n",
       "2         0.0  \n",
       "3         0.0  \n",
       "4         0.0  \n",
       "\n",
       "[5 rows x 20532 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "2e0ec7df",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(801, 20532)"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "3b7cf766",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>Class</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>sample_0</td>\n",
       "      <td>PRAD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>sample_1</td>\n",
       "      <td>LUAD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>sample_2</td>\n",
       "      <td>PRAD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>sample_3</td>\n",
       "      <td>PRAD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>sample_4</td>\n",
       "      <td>BRCA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Unnamed: 0 Class\n",
       "0   sample_0  PRAD\n",
       "1   sample_1  LUAD\n",
       "2   sample_2  PRAD\n",
       "3   sample_3  PRAD\n",
       "4   sample_4  BRCA"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "labels.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "0c3ea38e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(801, 2)"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "labels.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f0a97b63",
   "metadata": {},
   "source": [
    "Check for missing values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "f7309c81",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Unnamed: 0    0\n",
       "gene_13794    0\n",
       "gene_13692    0\n",
       "gene_13691    0\n",
       "gene_13690    0\n",
       "             ..\n",
       "gene_6841     0\n",
       "gene_6840     0\n",
       "gene_6839     0\n",
       "gene_6838     0\n",
       "gene_20530    0\n",
       "Length: 20532, dtype: int64"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.isnull().sum().sort_values(ascending = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "90f7f125",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Unnamed: 0    0\n",
       "Class         0\n",
       "dtype: int64"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "labels.isnull().sum().sort_values(ascending = False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a33f7bb6",
   "metadata": {},
   "source": [
    "# Exploratory Data Visualization"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "856e4195",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Class\n",
       "BRCA    300\n",
       "KIRC    146\n",
       "LUAD    141\n",
       "PRAD    136\n",
       "COAD     78\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "labels.value_counts('Class')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "84e979c3",
   "metadata": {},
   "source": [
    "Create a new column that has the cancer class labels 'Class'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "d6e6c098",
   "metadata": {},
   "outputs": [],
   "source": [
    "merged_df = pd.merge(df, labels, on='Unnamed: 0', how='left')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "577640bf",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>gene_0</th>\n",
       "      <th>gene_1</th>\n",
       "      <th>gene_2</th>\n",
       "      <th>gene_3</th>\n",
       "      <th>gene_4</th>\n",
       "      <th>gene_5</th>\n",
       "      <th>gene_6</th>\n",
       "      <th>gene_7</th>\n",
       "      <th>gene_8</th>\n",
       "      <th>...</th>\n",
       "      <th>gene_20522</th>\n",
       "      <th>gene_20523</th>\n",
       "      <th>gene_20524</th>\n",
       "      <th>gene_20525</th>\n",
       "      <th>gene_20526</th>\n",
       "      <th>gene_20527</th>\n",
       "      <th>gene_20528</th>\n",
       "      <th>gene_20529</th>\n",
       "      <th>gene_20530</th>\n",
       "      <th>Class</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>sample_0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.017209</td>\n",
       "      <td>3.265527</td>\n",
       "      <td>5.478487</td>\n",
       "      <td>10.431999</td>\n",
       "      <td>0.0</td>\n",
       "      <td>7.175175</td>\n",
       "      <td>0.591871</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>8.210257</td>\n",
       "      <td>9.723516</td>\n",
       "      <td>7.220030</td>\n",
       "      <td>9.119813</td>\n",
       "      <td>12.003135</td>\n",
       "      <td>9.650743</td>\n",
       "      <td>8.921326</td>\n",
       "      <td>5.286759</td>\n",
       "      <td>0.0</td>\n",
       "      <td>PRAD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>sample_1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.592732</td>\n",
       "      <td>1.588421</td>\n",
       "      <td>7.586157</td>\n",
       "      <td>9.623011</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.816049</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>7.323865</td>\n",
       "      <td>9.740931</td>\n",
       "      <td>6.256586</td>\n",
       "      <td>8.381612</td>\n",
       "      <td>12.674552</td>\n",
       "      <td>10.517059</td>\n",
       "      <td>9.397854</td>\n",
       "      <td>2.094168</td>\n",
       "      <td>0.0</td>\n",
       "      <td>LUAD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>sample_2</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.511759</td>\n",
       "      <td>4.327199</td>\n",
       "      <td>6.881787</td>\n",
       "      <td>9.870730</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.972130</td>\n",
       "      <td>0.452595</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>8.127123</td>\n",
       "      <td>10.908640</td>\n",
       "      <td>5.401607</td>\n",
       "      <td>9.911597</td>\n",
       "      <td>9.045255</td>\n",
       "      <td>9.788359</td>\n",
       "      <td>10.090470</td>\n",
       "      <td>1.683023</td>\n",
       "      <td>0.0</td>\n",
       "      <td>PRAD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>sample_3</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.663618</td>\n",
       "      <td>4.507649</td>\n",
       "      <td>6.659068</td>\n",
       "      <td>10.196184</td>\n",
       "      <td>0.0</td>\n",
       "      <td>7.843375</td>\n",
       "      <td>0.434882</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>8.792959</td>\n",
       "      <td>10.141520</td>\n",
       "      <td>8.942805</td>\n",
       "      <td>9.601208</td>\n",
       "      <td>11.392682</td>\n",
       "      <td>9.694814</td>\n",
       "      <td>9.684365</td>\n",
       "      <td>3.292001</td>\n",
       "      <td>0.0</td>\n",
       "      <td>PRAD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>sample_4</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.655741</td>\n",
       "      <td>2.821547</td>\n",
       "      <td>6.539454</td>\n",
       "      <td>9.738265</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.566967</td>\n",
       "      <td>0.360982</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>8.891425</td>\n",
       "      <td>10.373790</td>\n",
       "      <td>7.181162</td>\n",
       "      <td>9.846910</td>\n",
       "      <td>11.922439</td>\n",
       "      <td>9.217749</td>\n",
       "      <td>9.461191</td>\n",
       "      <td>5.110372</td>\n",
       "      <td>0.0</td>\n",
       "      <td>BRCA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 20533 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  Unnamed: 0  gene_0    gene_1    gene_2    gene_3     gene_4  gene_5  \\\n",
       "0   sample_0     0.0  2.017209  3.265527  5.478487  10.431999     0.0   \n",
       "1   sample_1     0.0  0.592732  1.588421  7.586157   9.623011     0.0   \n",
       "2   sample_2     0.0  3.511759  4.327199  6.881787   9.870730     0.0   \n",
       "3   sample_3     0.0  3.663618  4.507649  6.659068  10.196184     0.0   \n",
       "4   sample_4     0.0  2.655741  2.821547  6.539454   9.738265     0.0   \n",
       "\n",
       "     gene_6    gene_7  gene_8  ...  gene_20522  gene_20523  gene_20524  \\\n",
       "0  7.175175  0.591871     0.0  ...    8.210257    9.723516    7.220030   \n",
       "1  6.816049  0.000000     0.0  ...    7.323865    9.740931    6.256586   \n",
       "2  6.972130  0.452595     0.0  ...    8.127123   10.908640    5.401607   \n",
       "3  7.843375  0.434882     0.0  ...    8.792959   10.141520    8.942805   \n",
       "4  6.566967  0.360982     0.0  ...    8.891425   10.373790    7.181162   \n",
       "\n",
       "   gene_20525  gene_20526  gene_20527  gene_20528  gene_20529  gene_20530  \\\n",
       "0    9.119813   12.003135    9.650743    8.921326    5.286759         0.0   \n",
       "1    8.381612   12.674552   10.517059    9.397854    2.094168         0.0   \n",
       "2    9.911597    9.045255    9.788359   10.090470    1.683023         0.0   \n",
       "3    9.601208   11.392682    9.694814    9.684365    3.292001         0.0   \n",
       "4    9.846910   11.922439    9.217749    9.461191    5.110372         0.0   \n",
       "\n",
       "   Class  \n",
       "0   PRAD  \n",
       "1   LUAD  \n",
       "2   PRAD  \n",
       "3   PRAD  \n",
       "4   BRCA  \n",
       "\n",
       "[5 rows x 20533 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "2b9483a0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Class\n",
      "BRCA    300\n",
      "KIRC    146\n",
      "LUAD    141\n",
      "PRAD    136\n",
      "COAD     78\n",
      "Name: count, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "print(merged_df['Class'].value_counts())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "67d4c950",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<Axes: xlabel='Class'>"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAigAAAHLCAYAAADiNWM0AAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjcuMiwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8pXeV/AAAACXBIWXMAAA9hAAAPYQGoP6dpAAAvaklEQVR4nO3de1TU9b7/8dcAgqiAIMpAklJhVmhn5y2xtnhDWam51a3p9lZ2VdmH1LxUFrVS23bCLqa73KampZ5MM9OT0lYpt8e8lHczK6/JSCJyMQTF7++PTvPbk1iRMN+PzPOx1qwV8/kwvKcpefqdme84LMuyBAAAYBA/uwcAAAD4OQIFAAAYh0ABAADGIVAAAIBxCBQAAGAcAgUAABiHQAEAAMYJsHuA3+PixYs6ceKEQkJC5HA47B4HAAD8BpZlqbCwUDExMfLz++VjJFdloJw4cUKxsbF2jwEAAH6HY8eOqWHDhr+456oMlJCQEEk/3sHQ0FCbpwEAAL9FQUGBYmNj3b/Hf8lVGSg/Pa0TGhpKoAAAcJX5LS/P4EWyAADAOAQKAAAwDoECAACMQ6AAAADjECgAAMA4BAoAADAOgQIAAIxDoAAAAOMQKAAAwDgECgAAMA6BAgAAjFOhQJk1a5aaN2/u/gyctm3b6n/+53/c65ZlKT09XTExMQoODlZSUpL27t3rcRslJSVKTU1VZGSkateurZ49e+r48eOVc28AAEC1UKFAadiwoZ5//nlt27ZN27ZtU8eOHXX33Xe7I2TatGnKyMjQjBkztHXrVjmdTnXp0kWFhYXu20hLS9Py5cu1ePFibdy4UUVFRerevbvKysoq954BAICrlsOyLOtKbiAiIkIvvPCC7rvvPsXExCgtLU3jx4+X9OPRkqioKP3tb3/TQw89pPz8fNWvX18LFixQ//79JUknTpxQbGysVq9era5du/6mn1lQUKCwsDDl5+fzacYAAFwlKvL7+3e/BqWsrEyLFy/W2bNn1bZtWx06dEgul0vJycnuPUFBQWrfvr02bdokSdq+fbvOnz/vsScmJkYJCQnuPeUpKSlRQUGBxwUAAFRfARX9ht27d6tt27Y6d+6c6tSpo+XLl+vmm292B0ZUVJTH/qioKB05ckSS5HK5FBgYqPDw8Ev2uFyuy/7MqVOn6plnnqnoqFes8YRVXv+ZVeHw83fZPQIAABVS4SMoN954o3bs2KHNmzfrkUce0dChQ7Vv3z73usPh8NhvWdYl1/3cr+2ZOHGi8vPz3Zdjx45VdGwAAHAVqXCgBAYG6oYbblDLli01depU3XrrrXr55ZfldDol6ZIjITk5Oe6jKk6nU6WlpcrLy7vsnvIEBQW53zn00wUAAFRfV3weFMuyVFJSori4ODmdTmVmZrrXSktLlZWVpcTERElSixYtVKNGDY892dnZ2rNnj3sPAABAhV6D8vjjjyslJUWxsbEqLCzU4sWLtWHDBn300UdyOBxKS0vTlClTFB8fr/j4eE2ZMkW1atXSwIEDJUlhYWEaPny4xowZo3r16ikiIkJjx45Vs2bN1Llz5yq5gwAA4OpToUA5efKkBg8erOzsbIWFhal58+b66KOP1KVLF0nSuHHjVFxcrBEjRigvL09t2rTR2rVrFRIS4r6N6dOnKyAgQP369VNxcbE6deqkefPmyd/fv3LvGQAAuGpd8XlQ7OCt86DwLh4AACqPV86DAgAAUFUIFAAAYBwCBQAAGIdAAQAAxiFQAACAcQgUAABgHAIFAAAYh0ABAADGIVAAAIBxCBQAAGAcAgUAABiHQAEAAMYhUAAAgHEIFAAAYBwCBQAAGIdAAQAAxiFQAACAcQgUAABgHAIFAAAYh0ABAADGIVAAAIBxCBQAAGAcAgUAABiHQAEAAMYhUAAAgHEIFAAAYBwCBQAAGIdAAQAAxiFQAACAcQgUAABgHAIFAAAYh0ABAADGIVAAAIBxCBQAAGAcAgUAABiHQAEAAMYhUAAAgHEIFAAAYBwCBQAAGIdAAQAAxiFQAACAcQgUAABgHAIFAAAYh0ABAADGIVAAAIBxCBQAAGAcAgUAABiHQAEAAMapUKBMnTpVrVq1UkhIiBo0aKBevXrpwIEDHnuGDRsmh8Phcbn99ts99pSUlCg1NVWRkZGqXbu2evbsqePHj1/5vQEAANVChQIlKytLI0eO1ObNm5WZmakLFy4oOTlZZ8+e9djXrVs3ZWdnuy+rV6/2WE9LS9Py5cu1ePFibdy4UUVFRerevbvKysqu/B4BAICrXkBFNn/00UceX8+dO1cNGjTQ9u3b9cc//tF9fVBQkJxOZ7m3kZ+frzlz5mjBggXq3LmzJGnhwoWKjY3Vxx9/rK5du1b0PgAAgGrmil6Dkp+fL0mKiIjwuH7Dhg1q0KCBmjRpogceeEA5OTnute3bt+v8+fNKTk52XxcTE6OEhARt2rSp3J9TUlKigoICjwsAAKi+fnegWJal0aNH64477lBCQoL7+pSUFL399ttat26dXnzxRW3dulUdO3ZUSUmJJMnlcikwMFDh4eEetxcVFSWXy1Xuz5o6darCwsLcl9jY2N87NgAAuApU6Cmefzdq1Cjt2rVLGzdu9Li+f//+7n9OSEhQy5Yt1ahRI61atUq9e/e+7O1ZliWHw1Hu2sSJEzV69Gj31wUFBUQKAADV2O86gpKamqoPPvhA69evV8OGDX9xb3R0tBo1aqSDBw9KkpxOp0pLS5WXl+exLycnR1FRUeXeRlBQkEJDQz0uAACg+qpQoFiWpVGjRmnZsmVat26d4uLifvV7cnNzdezYMUVHR0uSWrRooRo1aigzM9O9Jzs7W3v27FFiYmIFxwcAANVRhZ7iGTlypN555x2tWLFCISEh7teMhIWFKTg4WEVFRUpPT1efPn0UHR2tw4cP6/HHH1dkZKT+9Kc/ufcOHz5cY8aMUb169RQREaGxY8eqWbNm7nf1AAAA31ahQJk1a5YkKSkpyeP6uXPnatiwYfL399fu3bv11ltv6cyZM4qOjlaHDh20ZMkShYSEuPdPnz5dAQEB6tevn4qLi9WpUyfNmzdP/v7+V36PAADAVc9hWZZl9xAVVVBQoLCwMOXn51fp61EaT1hVZbftTYefv8vuEQAAqNDvbz6LBwAAGIdAAQAAxiFQAACAcQgUAABgHAIFAAAYh0ABAADGIVAAAIBxCBQAAGAcAgUAABiHQAEAAMYhUAAAgHEIFAAAYBwCBQAAGIdAAQAAxiFQAACAcQgUAABgHAIFAAAYh0ABAADGIVAAAIBxCBQAAGAcAgUAABiHQAEAAMYhUAAAgHEIFAAAYBwCBQAAGIdAAQAAxiFQAACAcQgUAABgHAIFAAAYh0ABAADGIVAAAIBxCBQAAGAcAgUAABiHQAEAAMYhUAAAgHEIFAAAYBwCBQAAGIdAAQAAxiFQAACAcQgUAABgHAIFAAAYh0ABAADGIVAAAIBxCBQAAGAcAgUAABiHQAEAAMYhUAAAgHEqFChTp05Vq1atFBISogYNGqhXr146cOCAxx7LspSenq6YmBgFBwcrKSlJe/fu9dhTUlKi1NRURUZGqnbt2urZs6eOHz9+5fcGAABUCxUKlKysLI0cOVKbN29WZmamLly4oOTkZJ09e9a9Z9q0acrIyNCMGTO0detWOZ1OdenSRYWFhe49aWlpWr58uRYvXqyNGzeqqKhI3bt3V1lZWeXdMwAAcNVyWJZl/d5v/v7779WgQQNlZWXpj3/8oyzLUkxMjNLS0jR+/HhJPx4tiYqK0t/+9jc99NBDys/PV/369bVgwQL1799fknTixAnFxsZq9erV6tq166/+3IKCAoWFhSk/P1+hoaG/d/xf1XjCqiq7bW86/Pxddo8AAECFfn9f0WtQ8vPzJUkRERGSpEOHDsnlcik5Odm9JygoSO3bt9emTZskSdu3b9f58+c99sTExCghIcG95+dKSkpUUFDgcQEAANXX7w4Uy7I0evRo3XHHHUpISJAkuVwuSVJUVJTH3qioKPeay+VSYGCgwsPDL7vn56ZOnaqwsDD3JTY29veODQAArgK/O1BGjRqlXbt2adGiRZesORwOj68ty7rkup/7pT0TJ05Ufn6++3Ls2LHfOzYAALgK/K5ASU1N1QcffKD169erYcOG7uudTqckXXIkJCcnx31Uxel0qrS0VHl5eZfd83NBQUEKDQ31uAAAgOqrQoFiWZZGjRqlZcuWad26dYqLi/NYj4uLk9PpVGZmpvu60tJSZWVlKTExUZLUokUL1ahRw2NPdna29uzZ494DAAB8W0BFNo8cOVLvvPOOVqxYoZCQEPeRkrCwMAUHB8vhcCgtLU1TpkxRfHy84uPjNWXKFNWqVUsDBw507x0+fLjGjBmjevXqKSIiQmPHjlWzZs3UuXPnyr+HAADgqlOhQJk1a5YkKSkpyeP6uXPnatiwYZKkcePGqbi4WCNGjFBeXp7atGmjtWvXKiQkxL1/+vTpCggIUL9+/VRcXKxOnTpp3rx58vf3v7J7AwAAqoUrOg+KXTgPSsVwHhQAgAm8dh4UAACAqkCgAAAA4xAoAADAOAQKAAAwDoECAACMQ6AAAADjECgAAMA4BAoAADAOgQIAAIxDoAAAAOMQKAAAwDgECgAAMA6BAgAAjEOgAAAA4xAoAADAOAQKAAAwDoECAACMQ6AAAADjECgAAMA4BAoAADAOgQIAAIxDoAAAAOMQKAAAwDgECgAAMA6BAgAAjEOgAAAA4xAoAADAOAQKAAAwDoECAACMQ6AAAADjECgAAMA4BAoAADAOgQIAAIxDoAAAAOMQKAAAwDgECgAAMA6BAgAAjEOgAAAA4xAoAADAOAQKAAAwDoECAACMQ6AAAADjECgAAMA4BAoAADAOgQIAAIxDoAAAAOMQKAAAwDgECgAAME6FA+WTTz5Rjx49FBMTI4fDoffff99jfdiwYXI4HB6X22+/3WNPSUmJUlNTFRkZqdq1a6tnz546fvz4Fd0RAABQfVQ4UM6ePatbb71VM2bMuOyebt26KTs7231ZvXq1x3paWpqWL1+uxYsXa+PGjSoqKlL37t1VVlZW8XsAAACqnYCKfkNKSopSUlJ+cU9QUJCcTme5a/n5+ZozZ44WLFigzp07S5IWLlyo2NhYffzxx+ratWtFRwIAANVMlbwGZcOGDWrQoIGaNGmiBx54QDk5Oe617du36/z580pOTnZfFxMTo4SEBG3atKnc2yspKVFBQYHHBQAAVF+VHigpKSl6++23tW7dOr344ovaunWrOnbsqJKSEkmSy+VSYGCgwsPDPb4vKipKLper3NucOnWqwsLC3JfY2NjKHhsAABikwk/x/Jr+/fu7/zkhIUEtW7ZUo0aNtGrVKvXu3fuy32dZlhwOR7lrEydO1OjRo91fFxQUECkAAFRjVf424+joaDVq1EgHDx6UJDmdTpWWliovL89jX05OjqKiosq9jaCgIIWGhnpcAABA9VXlgZKbm6tjx44pOjpaktSiRQvVqFFDmZmZ7j3Z2dnas2ePEhMTq3ocAABwFajwUzxFRUX6+uuv3V8fOnRIO3bsUEREhCIiIpSenq4+ffooOjpahw8f1uOPP67IyEj96U9/kiSFhYVp+PDhGjNmjOrVq6eIiAiNHTtWzZo1c7+rBwAA+LYKB8q2bdvUoUMH99c/vTZk6NChmjVrlnbv3q233npLZ86cUXR0tDp06KAlS5YoJCTE/T3Tp09XQECA+vXrp+LiYnXq1Enz5s2Tv79/JdwlVEeNJ6yye4RKcfj5u+weAQCuChUOlKSkJFmWddn1NWvW/Opt1KxZU6+++qpeffXViv54AADgA/gsHgAAYBwCBQAAGIdAAQAAxiFQAACAcSr9TLIAqr/q8K4q3lEFmI0jKAAAwDgECgAAMA6BAgAAjEOgAAAA4xAoAADAOLyLBwCuYryjCtUVR1AAAIBxCBQAAGAcAgUAABiHQAEAAMYhUAAAgHEIFAAAYBwCBQAAGIdAAQAAxiFQAACAcQgUAABgHAIFAAAYh0ABAADGIVAAAIBxCBQAAGAcAgUAABiHQAEAAMYhUAAAgHEIFAAAYBwCBQAAGIdAAQAAxiFQAACAcQgUAABgHAIFAAAYh0ABAADGIVAAAIBxCBQAAGAcAgUAABiHQAEAAMYhUAAAgHEIFAAAYBwCBQAAGIdAAQAAxiFQAACAcQgUAABgHAIFAAAYh0ABAADGIVAAAIBxKhwon3zyiXr06KGYmBg5HA69//77HuuWZSk9PV0xMTEKDg5WUlKS9u7d67GnpKREqampioyMVO3atdWzZ08dP378iu4IAACoPiocKGfPntWtt96qGTNmlLs+bdo0ZWRkaMaMGdq6daucTqe6dOmiwsJC9560tDQtX75cixcv1saNG1VUVKTu3burrKzs998TAABQbQRU9BtSUlKUkpJS7pplWXrppZf0xBNPqHfv3pKk+fPnKyoqSu+8844eeugh5efna86cOVqwYIE6d+4sSVq4cKFiY2P18ccfq2vXrldwdwAAQHVQqa9BOXTokFwul5KTk93XBQUFqX379tq0aZMkafv27Tp//rzHnpiYGCUkJLj3/FxJSYkKCgo8LgAAoPqq1EBxuVySpKioKI/ro6Ki3Gsul0uBgYEKDw+/7J6fmzp1qsLCwtyX2NjYyhwbAAAYpkrexeNwODy+tizrkut+7pf2TJw4Ufn5+e7LsWPHKm1WAABgnkoNFKfTKUmXHAnJyclxH1VxOp0qLS1VXl7eZff8XFBQkEJDQz0uAACg+qrUQImLi5PT6VRmZqb7utLSUmVlZSkxMVGS1KJFC9WoUcNjT3Z2tvbs2ePeAwAAfFuF38VTVFSkr7/+2v31oUOHtGPHDkVEROjaa69VWlqapkyZovj4eMXHx2vKlCmqVauWBg4cKEkKCwvT8OHDNWbMGNWrV08REREaO3asmjVr5n5XDwAA8G0VDpRt27apQ4cO7q9Hjx4tSRo6dKjmzZuncePGqbi4WCNGjFBeXp7atGmjtWvXKiQkxP0906dPV0BAgPr166fi4mJ16tRJ8+bNk7+/fyXcJQAAcLWrcKAkJSXJsqzLrjscDqWnpys9Pf2ye2rWrKlXX31Vr776akV/PAAA8AF8Fg8AADAOgQIAAIxDoAAAAOMQKAAAwDgECgAAMA6BAgAAjEOgAAAA4xAoAADAOAQKAAAwDoECAACMQ6AAAADjECgAAMA4BAoAADAOgQIAAIxDoAAAAOMQKAAAwDgECgAAMA6BAgAAjEOgAAAA4wTYPQAAANVB4wmr7B6hUhx+/i67R5DEERQAAGAgAgUAABiHQAEAAMYhUAAAgHEIFAAAYBwCBQAAGIdAAQAAxiFQAACAcQgUAABgHAIFAAAYh0ABAADGIVAAAIBxCBQAAGAcAgUAABiHQAEAAMYhUAAAgHEIFAAAYBwCBQAAGIdAAQAAxiFQAACAcQgUAABgHAIFAAAYh0ABAADGIVAAAIBxCBQAAGAcAgUAABiHQAEAAMap9EBJT0+Xw+HwuDidTve6ZVlKT09XTEyMgoODlZSUpL1791b2GAAA4CpWJUdQbrnlFmVnZ7svu3fvdq9NmzZNGRkZmjFjhrZu3Sqn06kuXbqosLCwKkYBAABXoSoJlICAADmdTvelfv36kn48evLSSy/piSeeUO/evZWQkKD58+frhx9+0DvvvFMVowAAgKtQlQTKwYMHFRMTo7i4ON1zzz369ttvJUmHDh2Sy+VScnKye29QUJDat2+vTZs2Xfb2SkpKVFBQ4HEBAADVV6UHSps2bfTWW29pzZo1mj17tlwulxITE5WbmyuXyyVJioqK8vieqKgo91p5pk6dqrCwMPclNja2sscGAAAGqfRASUlJUZ8+fdSsWTN17txZq1atkiTNnz/fvcfhcHh8j2VZl1z37yZOnKj8/Hz35dixY5U9NgAAMEiVv824du3aatasmQ4ePOh+N8/Pj5bk5ORcclTl3wUFBSk0NNTjAgAAqq8qD5SSkhLt379f0dHRiouLk9PpVGZmpnu9tLRUWVlZSkxMrOpRAADAVSKgsm9w7Nix6tGjh6699lrl5OToueeeU0FBgYYOHSqHw6G0tDRNmTJF8fHxio+P15QpU1SrVi0NHDiwskcBAABXqUoPlOPHj2vAgAE6deqU6tevr9tvv12bN29Wo0aNJEnjxo1TcXGxRowYoby8PLVp00Zr165VSEhIZY8CAACuUpUeKIsXL/7FdYfDofT0dKWnp1f2jwYAANUEn8UDAACMQ6AAAADjECgAAMA4BAoAADAOgQIAAIxDoAAAAOMQKAAAwDgECgAAMA6BAgAAjEOgAAAA4xAoAADAOAQKAAAwDoECAACMQ6AAAADjECgAAMA4BAoAADAOgQIAAIxDoAAAAOMQKAAAwDgECgAAMA6BAgAAjEOgAAAA4xAoAADAOAQKAAAwDoECAACMQ6AAAADjECgAAMA4BAoAADAOgQIAAIxDoAAAAOMQKAAAwDgECgAAMA6BAgAAjEOgAAAA4xAoAADAOAQKAAAwDoECAACMQ6AAAADjECgAAMA4BAoAADAOgQIAAIxDoAAAAOMQKAAAwDgECgAAMA6BAgAAjEOgAAAA4xAoAADAOAQKAAAwjq2BMnPmTMXFxalmzZpq0aKFPv30UzvHAQAAhrAtUJYsWaK0tDQ98cQT+uKLL3TnnXcqJSVFR48etWskAABgCNsCJSMjQ8OHD9f999+vm266SS+99JJiY2M1a9Ysu0YCAACGCLDjh5aWlmr79u2aMGGCx/XJycnatGnTJftLSkpUUlLi/jo/P1+SVFBQUKVzXiz5oUpv31uq+t+TN/BYmKU6PB48FubgsTBLVT4eP922ZVm/uteWQDl16pTKysoUFRXlcX1UVJRcLtcl+6dOnapnnnnmkutjY2OrbMbqJOwluyfAT3gszMFjYQ4eC7N44/EoLCxUWFjYL+6xJVB+4nA4PL62LOuS6yRp4sSJGj16tPvrixcv6vTp06pXr165+68WBQUFio2N1bFjxxQaGmr3OD6Nx8IcPBZm4fEwR3V4LCzLUmFhoWJiYn51ry2BEhkZKX9//0uOluTk5FxyVEWSgoKCFBQU5HFd3bp1q3JErwoNDb1q/2OrbngszMFjYRYeD3Nc7Y/Frx05+YktL5INDAxUixYtlJmZ6XF9ZmamEhMT7RgJAAAYxLaneEaPHq3BgwerZcuWatu2rd544w0dPXpUDz/8sF0jAQAAQ9gWKP3791dubq6effZZZWdnKyEhQatXr1ajRo3sGsnrgoKC9PTTT1/y9BW8j8fCHDwWZuHxMIevPRYO67e81wcAAMCL+CweAABgHAIFAAAYh0ABAADGIVAAAIBxCBQAAGAcAgUAABjH1s/iwf+3e/duzZkzRy+99JLdo/iMgoIC1alTR35+np1eVlams2fPXtWnkgaulGVZys3NlcPhUL169ewexyetW7dOy5Yt0+HDh+VwOBQXF6e+ffvqj3/8o92jeQVHUGxUUFCg119/Xa1bt9att96qDRs22D2Sz1i+fLlatmypc+fOXbJWUlKiVq1aaeXKlTZM5rsOHjyo//qv/9KoUaOUmpqqjIwMffvtt3aP5XNcLpeGDBmi8PBwRUVFqUGDBgoPD9d9992nkydP2j2ez3j44YfVuXNnLVq0SLm5ufr+++/19ttvq0OHDkpNTbV7PO+w4HUbNmywBg8ebNWqVcvy8/Ozxo8fbx08eNDusXxKly5drNmzZ192fc6cOVZycrIXJ/JtU6ZMsQICAiw/Pz/L6XRaUVFRlp+fn1WjRg3rhRdesHs8n5Gfn2/FxcVZ9evXt9LS0qy///3v1qxZs6zU1FQrMjLSio+PtwoLC+0es9pbtmyZFRgYaM2dO9e6ePGi+/qysjJrzpw5VmBgoLVixQobJ/QOAsVLTpw4YU2ePNm6/vrrLafTaT366KPW1q1brYCAAGvv3r12j+dzoqOjfzEKDx48aEVHR3txIt+1bt06y8/Pz3r66aet06dPu6/Pzc21Jk2aZPn7+1tZWVk2Tug7nn32WeuGG26wcnJyLlk7efKkdcMNN1iTJ0+2YTLf0qNHD2vChAmXXR83bpzVs2dPL05kD0517yU1a9bUn//8Zw0aNEhdunRxv+6hRo0a2rlzp26++WabJ/QtwcHB+uKLL9S0adNy1/fv36/bbrtNxcXFXp7M9/Tv319169bV66+/Xu76gw8+qMLCQi1atMjLk/me22+/XQ899JDuvffectfffPNNzZ49W//7v//r5cl8S8OGDbVs2TK1bt263PUtW7aod+/eOn78uJcn8y5eg+IljRo10saNG/XJJ5/oq6++snscn9e4cWNt27btsuvbtm3zqQ+utNOWLVs0ePDgy64PHjxYmzdv9uJEvuurr75SYmLiZdcTExN14MABL07km06dOqVrrrnmsuvXXHONcnNzvTiRPQgULzlw4IAWLlyo7OxstWrVSi1atND06dMlSQ6Hw+bpfE/v3r31xBNPlPuiP5fLpSeffFJ9+vSxYTLfc/LkSTVu3Piy63FxcXK5XN4byIcVFBSobt26l12vW7euCgoKvDeQjyotLVVgYOBl1wMCAlRaWurFiezBUzw2KCoq0qJFi/Tmm2/qs88+U/v27TVw4ED16tVL9evXt3s8n1BYWKi2bdvq6NGjGjRokG688UY5HA7t379fb7/9tmJjY7V582aFhITYPWq15+fnJ5fLpQYNGpS7fvLkScXExKisrMzLk/kef39/uVyuy/45xGPhHX5+fnrwwQdVq1atctd/+OEHzZ49u9o/DgSKzfbv3685c+ZowYIFOn36tM6fP2/3SD4jPz9fEydO1JIlS5SXlydJCg8PV//+/TVlypRf/JskKo+fn5+ee+451alTp9z1wsJCPfXUU9X+D2MT+Pn5KSws7LJHdS3LUkFBAY9FFUtKSvpNR9bXr1/vhWnsQ6AY4sKFC/rggw/Uu3dvu0fxOZZl6dSpU7IsS/Xr13f/wbB161a1atXK5umqv8aNG/+mP4wPHTrkhWl82/z583/TvqFDh1bxJACB4jUnTpxQRkaGnnrqqUvOUJqfn6/nnntOY8eOVVRUlE0T+qaioiL5+/srODjYfd2OHTs0adIkrV69mr8pAj9z4cIFBQRwEnI7+cqZx3mRrJdkZGSooKCg3NOnh4WFqbCwUBkZGTZM5puOHz+udu3aKSwsTGFhYRo9erR++OEHDRkyRK1atVJQUJA2btxo95iQlJubW+3/IL4a7Nu3T2PGjPnFd5eg6vjimccJFC/56KOPNGTIkMuuDxkyRB9++KEXJ/JtEyZMUFFRkV5++WW1a9dOL7/8su68804FBAToq6++0tKlS9W2bVu7x/RZlmVpzZo16tevn2JiYjR58mS7R/JJRUVF+sc//qG2bduqefPm+uyzzzRhwgS7x/IpWVlZGjJkiKKjozVixAh17NhRX331lXbs2GH3aFXPjrPD+aJatWpZR44cuez6kSNHrFq1anlxIt8WExNjbdy40bIsy8rOzrYcDoc1depUm6fCoUOHrEmTJlmxsbGWn5+fNXjwYCszM9O6cOGC3aP5lE8//dQaOnSoVadOHatZs2aWv7+/+/8XVD3OPP4jjqB4SXBwsA4fPnzZ9cOHD3u8DgJVy+Vy6frrr5ckOZ1OBQcH6+6777Z5Kt9UUlKiRYsWqVOnTrrpppu0Z88eZWRkyM/PTxMmTFDnzp3l7+9v95g+Ydq0aWratKnuuece1a9fXxs3btSuXbvkcDgUHh5u93g+Iy4uTvv379drr72m7777ThkZGWrZsqXdY3kdr3TykjZt2mjBggWX/Zjst95667KnNUbV+Pdfen5+fqpZs6aN0/iua665RjfffLMGDRqkpUuXun8RDhgwwObJfM/jjz+u8ePH69lnnyUKbfTTmcevvfZaNWrU6LIfyVHdESheMnbsWHXp0kVhYWF67LHH3O/WOXnypKZNm6Z58+Zp7dq1Nk/pOyzLUqdOndzvRiguLlaPHj0uOXvj559/bsd4PqWsrEwOh0MOh4NfijZ79tlnNW/ePC1YsEADBgzQ4MGDlZCQYPdYPufAgQP617/+pTlz5qhVq1Zq0qSJBg0aJMm3zjzO24y96PXXX9d//ud/6vz58woNDZXD4VB+fr5q1Kih6dOn65FHHrF7RJ/xzDPP/KZ9Tz/9dBVPgnPnzum9997TnDlztHnzZqWkpGjQoEHq37+/duzYwQdp2iArK0tvvvmm3nvvPV1//fXau3evsrKy1K5dO7tH8zm+fOZxAsXLvvvuO/33f/+3vv76a1mWpSZNmqhv375q2LCh3aMBtvvmm280d+5czZ8/X999950GDBigYcOGqWPHjhxdsUFhYaHefvttzZ07V9u3b1fr1q3Vt29fjR492u7RfNK+ffs0Z84cLVy40CfOPE6gGOLzzz/XU089xVuNAUkXL17UmjVrNGfOHK1cuVIhISE6deqU3WP5tJ9ODvbOO+8oJyfH7nF82vnz57Vy5cpqf+ZxXoPiRZmZmVq7dq0CAwM1fPhwXXfddfryyy81YcIErVy5Ul26dLF7RJ/xhz/84Tc9l8trUOzh5+enlJQUpaSk6Pvvv9eCBQvsHsnnNWvWTOPGjdOFCxfsHsVnvPvuu1q0aJG++uorORwOxcfHa+DAgerbt2+1jxOJQPGa+fPn695771VERIROnz6t2bNnKyMjQyNGjFCfPn20c+dOXozmRb169frVPfn5+VU/CH5V/fr1eUrBi/bt26f169erRo0a6tevn+rWratTp05p8uTJev3119W4cWO7R6z2Ll68qAEDBujdd99VkyZN1LRpU1mWpb1796pfv37q16+fFi1aVP1fMGvbGVh8zK233uo+EdiSJUssh8Nh3XbbbdbXX39t82S+6YUXXvjF9fz8fKtNmzZemsa3ORwOy8/P75JL3bp1rTZt2ljvvfee3SP6jJUrV1qBgYGWw+GwHA6Hdf3111vr1q2zIiMjraSkJGvlypV2j+gTXnzxRSsiIqLcf98rVqywIiIirOnTp3t/MC/jNSheEhISol27dikuLk4XL15UUFCQPv74Y7Vv397u0XxScHCwZs6cqXvvvfeStaKiIiUnJ+vMmTPat2+fDdP5lhUrVpR7/ZkzZ7Rlyxb3i2b//Oc/e3ky39O2bVu1bt1akydP1htvvKGxY8cqPj5es2fPvuw5nFD5mjdvrrS0NN13333lrv/0QYG7d+/28mTeRaB4iZ+fn1wulxo0aCDpx2DZuXOnrrvuOpsn801Lly7VoEGDtHjxYo+ne36Kk9zcXGVlZcnpdNo3JCRJr732mt566y199tlndo9S7dWtW1dbtmxRkyZNdOHCBdWsWVMrV65USkqK3aP5lODgYB04cEDXXnttuetHjhxR06ZNVVxc7OXJvIvXoHjRmjVrFBYWJunH5xj/+c9/as+ePR57evbsacdoPqdv3746c+aMBg4cqFWrVqlDhw4qKipSt27ddOrUKeLEIMnJyXryySftHsMnFBQUqG7dupKkgIAABQcHq0mTJvYO5YOCg4N15syZywZKQUGBT3w0CoHiRUOHDvX4+qGHHvL42uFwqKyszJsj+bT7779fp0+fVq9evbRixQpNmjRJLpdLWVlZio6Otns8/J/i4mI+hsCL9u3bJ5fLJenHMy4fOHBAZ8+e9djTvHlzO0bzGW3bttWsWbM0a9asctdfe+01n/i0dQLFSy5evGj3CCjHuHHjlJeXp06dOqlx48bKysrSNddcY/dY+DdvvPGG/uM//sPuMXxGx44dPb7u3r27pB//AmVZFn+R8oInnnhCSUlJys3N1dixY93v4tm/f79efPFFrVixQuvXr7d7zCpHoHhZbm6u6tWrJ0k6duyYZs+erXPnzqlHjx668847bZ7Od/z8HAI1atRQZGSk/vrXv3pcv2zZMm+O5ZMu9xbi/Px8bdu2Td98840+/fRTL0/lmw4dOmT3CJCUmJioJUuW6MEHH9R7773nsRYeHq5Fixb5xMcO8CJZL9m9e7d69OihY8eOKT4+XosXL1a3bt109uxZ+fn56ezZs1q6dOlvOj8Hrlx5794pz9y5c6t4EnTo0KHc60NDQ9W0aVN1795dc+fO1ZtvvunlyXxPcXGxxo4dq/fff1/nz59X586d9corrygyMtLu0XzSDz/8oDVr1ujgwYOSpCZNmig5OVm1atWyeTLvIFC8JCUlRQEBARo/frwWLlyoDz/8UMnJyfrHP/4hSUpNTdX27du1efNmmycFzLJz507ddtttPK3gBY899phmzpypv/zlL6pZs6YWLVqkpKQkvfvuu3aP5lPWrVunUaNGafPmzQoNDfVYy8/PV2Jiov7+979X+6PuBIqXREZGat26dWrevLmKiooUGhqqLVu2qGXLlpKkL7/8UrfffrvOnDlj76CAYQgU77n++us1efJk3XPPPZKkLVu2qF27djp37hwf1uhFPXv2VIcOHfToo4+Wu/7KK69o/fr1Wr58uZcn8y4/uwfwFadPn3a/bbVOnTqqXbu2IiIi3Ovh4eEqLCy0azwA0LFjxzz+Vt66dWsFBAToxIkTNk7le3bu3Klu3bpddj05OVnbt2/34kT2IFC86Oefm1DtP0cBwFWlrKxMgYGBHtcFBATwAYFedvLkSdWoUeOy6wEBAfr++++9OJE9eBePFw0bNkxBQUGSpHPnzunhhx9W7dq1JUklJSV2jgbY5tc+lZWnPb3HsiyPP6ekS/+sknh3W1W75pprtHv3bt1www3lru/atcsnztXEa1C8hHeNAOXj/w1z8FiYITU1VRs2bNDWrVsvOUlhcXGxWrdurQ4dOuiVV16xaULvIFAAADDIyZMnddttt8nf31+jRo3SjTfeKIfDof379+u1115TWVmZPv/8c0VFRdk9apUiUAAAMMyRI0f0yCOPaM2aNfrp17TD4VDXrl01c+ZMNW7c2N4BvYBAAQDAUHl5efr6669lWZbi4+MVHh5u90heQ6AAAADj8DZjAABgHAIFAAAYh0ABAADGIVAAAIBxCBQAtnA4HHr//fftHgOAoQgUAFXC5XIpNTVV1113nYKCghQbG6sePXron//8p92jAbgK8Fk8ACrd4cOH1a5dO9WtW1fTpk1T8+bNdf78ea1Zs0YjR47Ul19+afeIAAzHERQAlW7EiBFyOBzasmWL+vbtqyZNmuiWW27R6NGjtXnz5nK/Z/z48WrSpIlq1aql6667TpMmTdL58+fd6zt37lSHDh0UEhKi0NBQtWjRQtu2bZP041k3e/ToofDwcNWuXVu33HKLVq9e7ZX7CqBqcAQFQKU6ffq0PvroI02ePNnjE3B/Urdu3XK/LyQkRPPmzVNMTIx2796tBx54QCEhIRo3bpwk6S9/+Yv+8Ic/aNasWfL399eOHTvcH0k/cuRIlZaW6pNPPlHt2rW1b98+1alTp8ruI4CqR6AAqFQ/nZa7adOmFfq+J5980v3PjRs31pgxY7RkyRJ3oBw9elSPPfaY+3bj4+Pd+48ePao+ffqoWbNmkqTrrrvuSu8GAJvxFA+ASvXvH2xWEUuXLtUdd9whp9OpOnXqaNKkSTp69Kh7ffTo0br//vvVuXNnPf/88/rmm2/ca3/961/13HPPqV27dnr66ae1a9euyrkzAGxDoACoVPHx8e6Phv+tNm/erHvuuUcpKSn68MMP9cUXX+iJJ55QaWmpe096err27t2ru+66S+vWrdPNN9+s5cuXS5Luv/9+ffvttxo8eLB2796tli1b6tVXX630+wbAe/iwQACVLiUlRbt379aBAwcueR3KmTNnVLduXTkcDi1fvly9evXSiy++qJkzZ3ocFbn//vu1dOlSnTlzptyfMWDAAJ09e1YffPDBJWsTJ07UqlWrOJICXMU4ggKg0s2cOVNlZWVq3bq13nvvPR08eFD79+/XK6+8orZt216y/4YbbtDRo0e1ePFiffPNN3rllVfcR0ckqbi4WKNGjdKGDRt05MgR/etf/9LWrVt10003SZLS0tK0Zs0aHTp0SJ9//rnWrVvnXgNwdeJFsgAqXVxcnD7//HNNnjxZY8aMUXZ2turXr68WLVpo1qxZl+y/++679eijj2rUqFEqKSnRXXfdpUmTJik9PV2S5O/vr9zcXA0ZMkQnT55UZGSkevfurWeeeUaSVFZWppEjR+r48eMKDQ1Vt27dNH36dG/eZQCVjKd4AACAcXiKBwAAGIdAAQAAxiFQAACAcQgUAABgHAIFAAAYh0ABAADGIVAAAIBxCBQAAGAcAgUAABiHQAEAAMYhUAAAgHH+H3TNy1GULQ48AAAAAElFTkSuQmCC",
      "text/plain": [
       "<Figure size 640x480 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "merged_df['Class'].value_counts().plot.bar()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "acb20e94",
   "metadata": {},
   "source": [
    "remove the sample ID 'Unnamed: 0' colum"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "881d68e9",
   "metadata": {},
   "outputs": [],
   "source": [
    "merged_df = merged_df.drop(columns=['Unnamed: 0'])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "eaa59f0f",
   "metadata": {},
   "source": [
    "# Data Processing"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "eec4fca8",
   "metadata": {},
   "source": [
    "We need to put the data into a format that is compatible with Scikitlearn.\n",
    "\n",
    "Seperate the feature values from the class. we do this because scikit-learn requires that features and class"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "28d677b5",
   "metadata": {},
   "outputs": [],
   "source": [
    "X=merged_df.iloc[:,0:-1]\n",
    "y=merged_df.iloc[:,-1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "ca627c66",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_0</th>\n",
       "      <th>gene_1</th>\n",
       "      <th>gene_2</th>\n",
       "      <th>gene_3</th>\n",
       "      <th>gene_4</th>\n",
       "      <th>gene_5</th>\n",
       "      <th>gene_6</th>\n",
       "      <th>gene_7</th>\n",
       "      <th>gene_8</th>\n",
       "      <th>gene_9</th>\n",
       "      <th>...</th>\n",
       "      <th>gene_20521</th>\n",
       "      <th>gene_20522</th>\n",
       "      <th>gene_20523</th>\n",
       "      <th>gene_20524</th>\n",
       "      <th>gene_20525</th>\n",
       "      <th>gene_20526</th>\n",
       "      <th>gene_20527</th>\n",
       "      <th>gene_20528</th>\n",
       "      <th>gene_20529</th>\n",
       "      <th>gene_20530</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.0</td>\n",
       "      <td>2.017209</td>\n",
       "      <td>3.265527</td>\n",
       "      <td>5.478487</td>\n",
       "      <td>10.431999</td>\n",
       "      <td>0.0</td>\n",
       "      <td>7.175175</td>\n",
       "      <td>0.591871</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>4.926711</td>\n",
       "      <td>8.210257</td>\n",
       "      <td>9.723516</td>\n",
       "      <td>7.220030</td>\n",
       "      <td>9.119813</td>\n",
       "      <td>12.003135</td>\n",
       "      <td>9.650743</td>\n",
       "      <td>8.921326</td>\n",
       "      <td>5.286759</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.592732</td>\n",
       "      <td>1.588421</td>\n",
       "      <td>7.586157</td>\n",
       "      <td>9.623011</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.816049</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>4.593372</td>\n",
       "      <td>7.323865</td>\n",
       "      <td>9.740931</td>\n",
       "      <td>6.256586</td>\n",
       "      <td>8.381612</td>\n",
       "      <td>12.674552</td>\n",
       "      <td>10.517059</td>\n",
       "      <td>9.397854</td>\n",
       "      <td>2.094168</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0.0</td>\n",
       "      <td>3.511759</td>\n",
       "      <td>4.327199</td>\n",
       "      <td>6.881787</td>\n",
       "      <td>9.870730</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.972130</td>\n",
       "      <td>0.452595</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>5.125213</td>\n",
       "      <td>8.127123</td>\n",
       "      <td>10.908640</td>\n",
       "      <td>5.401607</td>\n",
       "      <td>9.911597</td>\n",
       "      <td>9.045255</td>\n",
       "      <td>9.788359</td>\n",
       "      <td>10.090470</td>\n",
       "      <td>1.683023</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0.0</td>\n",
       "      <td>3.663618</td>\n",
       "      <td>4.507649</td>\n",
       "      <td>6.659068</td>\n",
       "      <td>10.196184</td>\n",
       "      <td>0.0</td>\n",
       "      <td>7.843375</td>\n",
       "      <td>0.434882</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>6.076566</td>\n",
       "      <td>8.792959</td>\n",
       "      <td>10.141520</td>\n",
       "      <td>8.942805</td>\n",
       "      <td>9.601208</td>\n",
       "      <td>11.392682</td>\n",
       "      <td>9.694814</td>\n",
       "      <td>9.684365</td>\n",
       "      <td>3.292001</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0.0</td>\n",
       "      <td>2.655741</td>\n",
       "      <td>2.821547</td>\n",
       "      <td>6.539454</td>\n",
       "      <td>9.738265</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.566967</td>\n",
       "      <td>0.360982</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>5.996032</td>\n",
       "      <td>8.891425</td>\n",
       "      <td>10.373790</td>\n",
       "      <td>7.181162</td>\n",
       "      <td>9.846910</td>\n",
       "      <td>11.922439</td>\n",
       "      <td>9.217749</td>\n",
       "      <td>9.461191</td>\n",
       "      <td>5.110372</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>796</th>\n",
       "      <td>0.0</td>\n",
       "      <td>1.865642</td>\n",
       "      <td>2.718197</td>\n",
       "      <td>7.350099</td>\n",
       "      <td>10.006003</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.764792</td>\n",
       "      <td>0.496922</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>6.088133</td>\n",
       "      <td>9.118313</td>\n",
       "      <td>10.004852</td>\n",
       "      <td>4.484415</td>\n",
       "      <td>9.614701</td>\n",
       "      <td>12.031267</td>\n",
       "      <td>9.813063</td>\n",
       "      <td>10.092770</td>\n",
       "      <td>8.819269</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>797</th>\n",
       "      <td>0.0</td>\n",
       "      <td>3.942955</td>\n",
       "      <td>4.453807</td>\n",
       "      <td>6.346597</td>\n",
       "      <td>10.056868</td>\n",
       "      <td>0.0</td>\n",
       "      <td>7.320331</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>6.371876</td>\n",
       "      <td>9.623335</td>\n",
       "      <td>9.823921</td>\n",
       "      <td>6.555327</td>\n",
       "      <td>9.064002</td>\n",
       "      <td>11.633422</td>\n",
       "      <td>10.317266</td>\n",
       "      <td>8.745983</td>\n",
       "      <td>9.659081</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>798</th>\n",
       "      <td>0.0</td>\n",
       "      <td>3.249582</td>\n",
       "      <td>3.707492</td>\n",
       "      <td>8.185901</td>\n",
       "      <td>9.504082</td>\n",
       "      <td>0.0</td>\n",
       "      <td>7.536589</td>\n",
       "      <td>1.811101</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>5.719386</td>\n",
       "      <td>8.610704</td>\n",
       "      <td>10.485517</td>\n",
       "      <td>3.589763</td>\n",
       "      <td>9.350636</td>\n",
       "      <td>12.180944</td>\n",
       "      <td>10.681194</td>\n",
       "      <td>9.466711</td>\n",
       "      <td>4.677458</td>\n",
       "      <td>0.586693</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>799</th>\n",
       "      <td>0.0</td>\n",
       "      <td>2.590339</td>\n",
       "      <td>2.787976</td>\n",
       "      <td>7.318624</td>\n",
       "      <td>9.987136</td>\n",
       "      <td>0.0</td>\n",
       "      <td>9.213464</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>5.785237</td>\n",
       "      <td>8.605387</td>\n",
       "      <td>11.004677</td>\n",
       "      <td>4.745888</td>\n",
       "      <td>9.626383</td>\n",
       "      <td>11.198279</td>\n",
       "      <td>10.335513</td>\n",
       "      <td>10.400581</td>\n",
       "      <td>5.718751</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>800</th>\n",
       "      <td>0.0</td>\n",
       "      <td>2.325242</td>\n",
       "      <td>3.805932</td>\n",
       "      <td>6.530246</td>\n",
       "      <td>9.560367</td>\n",
       "      <td>0.0</td>\n",
       "      <td>7.957027</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>6.403075</td>\n",
       "      <td>8.594354</td>\n",
       "      <td>10.243079</td>\n",
       "      <td>9.139459</td>\n",
       "      <td>10.102934</td>\n",
       "      <td>11.641081</td>\n",
       "      <td>10.607358</td>\n",
       "      <td>9.844794</td>\n",
       "      <td>4.550716</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>801 rows Ã— 20531 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     gene_0    gene_1    gene_2    gene_3     gene_4  gene_5    gene_6  \\\n",
       "0       0.0  2.017209  3.265527  5.478487  10.431999     0.0  7.175175   \n",
       "1       0.0  0.592732  1.588421  7.586157   9.623011     0.0  6.816049   \n",
       "2       0.0  3.511759  4.327199  6.881787   9.870730     0.0  6.972130   \n",
       "3       0.0  3.663618  4.507649  6.659068  10.196184     0.0  7.843375   \n",
       "4       0.0  2.655741  2.821547  6.539454   9.738265     0.0  6.566967   \n",
       "..      ...       ...       ...       ...        ...     ...       ...   \n",
       "796     0.0  1.865642  2.718197  7.350099  10.006003     0.0  6.764792   \n",
       "797     0.0  3.942955  4.453807  6.346597  10.056868     0.0  7.320331   \n",
       "798     0.0  3.249582  3.707492  8.185901   9.504082     0.0  7.536589   \n",
       "799     0.0  2.590339  2.787976  7.318624   9.987136     0.0  9.213464   \n",
       "800     0.0  2.325242  3.805932  6.530246   9.560367     0.0  7.957027   \n",
       "\n",
       "       gene_7  gene_8  gene_9  ...  gene_20521  gene_20522  gene_20523  \\\n",
       "0    0.591871     0.0     0.0  ...    4.926711    8.210257    9.723516   \n",
       "1    0.000000     0.0     0.0  ...    4.593372    7.323865    9.740931   \n",
       "2    0.452595     0.0     0.0  ...    5.125213    8.127123   10.908640   \n",
       "3    0.434882     0.0     0.0  ...    6.076566    8.792959   10.141520   \n",
       "4    0.360982     0.0     0.0  ...    5.996032    8.891425   10.373790   \n",
       "..        ...     ...     ...  ...         ...         ...         ...   \n",
       "796  0.496922     0.0     0.0  ...    6.088133    9.118313   10.004852   \n",
       "797  0.000000     0.0     0.0  ...    6.371876    9.623335    9.823921   \n",
       "798  1.811101     0.0     0.0  ...    5.719386    8.610704   10.485517   \n",
       "799  0.000000     0.0     0.0  ...    5.785237    8.605387   11.004677   \n",
       "800  0.000000     0.0     0.0  ...    6.403075    8.594354   10.243079   \n",
       "\n",
       "     gene_20524  gene_20525  gene_20526  gene_20527  gene_20528  gene_20529  \\\n",
       "0      7.220030    9.119813   12.003135    9.650743    8.921326    5.286759   \n",
       "1      6.256586    8.381612   12.674552   10.517059    9.397854    2.094168   \n",
       "2      5.401607    9.911597    9.045255    9.788359   10.090470    1.683023   \n",
       "3      8.942805    9.601208   11.392682    9.694814    9.684365    3.292001   \n",
       "4      7.181162    9.846910   11.922439    9.217749    9.461191    5.110372   \n",
       "..          ...         ...         ...         ...         ...         ...   \n",
       "796    4.484415    9.614701   12.031267    9.813063   10.092770    8.819269   \n",
       "797    6.555327    9.064002   11.633422   10.317266    8.745983    9.659081   \n",
       "798    3.589763    9.350636   12.180944   10.681194    9.466711    4.677458   \n",
       "799    4.745888    9.626383   11.198279   10.335513   10.400581    5.718751   \n",
       "800    9.139459   10.102934   11.641081   10.607358    9.844794    4.550716   \n",
       "\n",
       "     gene_20530  \n",
       "0      0.000000  \n",
       "1      0.000000  \n",
       "2      0.000000  \n",
       "3      0.000000  \n",
       "4      0.000000  \n",
       "..          ...  \n",
       "796    0.000000  \n",
       "797    0.000000  \n",
       "798    0.586693  \n",
       "799    0.000000  \n",
       "800    0.000000  \n",
       "\n",
       "[801 rows x 20531 columns]"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "X"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "c5334742",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0      PRAD\n",
       "1      LUAD\n",
       "2      PRAD\n",
       "3      PRAD\n",
       "4      BRCA\n",
       "       ... \n",
       "796    BRCA\n",
       "797    LUAD\n",
       "798    COAD\n",
       "799    PRAD\n",
       "800    PRAD\n",
       "Name: Class, Length: 801, dtype: object"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "y"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b0243fa7",
   "metadata": {},
   "source": [
    "Encoding labels"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a53917c0",
   "metadata": {},
   "source": [
    "The labels for this data are categorical and we therefore have to convert them to numeric forms. This is referred to as encoding. Machine learning models usually require input data to be in numeric forms, hence we encoding the labels.\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "b1f3076b",
   "metadata": {},
   "outputs": [],
   "source": [
    "#let's encode target labels (y) with values between 0 and n_classes-1.\n",
    "#encoding will be done using the LabelEncoder\n",
    "label_encoder=LabelEncoder()\n",
    "label_encoder.fit(y)\n",
    "y_encoded=label_encoder.transform(y)\n",
    "labels=label_encoder.classes_\n",
    "classes=np.unique(y_encoded)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "1d27ec65",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['BRCA', 'COAD', 'KIRC', 'LUAD', 'PRAD'], dtype=object)"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "labels"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "93e20283",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([0, 1, 2, 3, 4])"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "classes"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "33c98fef",
   "metadata": {},
   "source": [
    "# Data Splitting"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "f70d11a2",
   "metadata": {},
   "outputs": [],
   "source": [
    "X_train,X_test,y_train,y_test=train_test_split(X, y_encoded, test_size=0.2,random_state=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "1554e60a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_0</th>\n",
       "      <th>gene_1</th>\n",
       "      <th>gene_2</th>\n",
       "      <th>gene_3</th>\n",
       "      <th>gene_4</th>\n",
       "      <th>gene_5</th>\n",
       "      <th>gene_6</th>\n",
       "      <th>gene_7</th>\n",
       "      <th>gene_8</th>\n",
       "      <th>gene_9</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>count</th>\n",
       "      <td>801.000000</td>\n",
       "      <td>801.000000</td>\n",
       "      <td>801.000000</td>\n",
       "      <td>801.000000</td>\n",
       "      <td>801.000000</td>\n",
       "      <td>801.0</td>\n",
       "      <td>801.000000</td>\n",
       "      <td>801.000000</td>\n",
       "      <td>801.000000</td>\n",
       "      <td>801.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mean</th>\n",
       "      <td>0.026642</td>\n",
       "      <td>3.010909</td>\n",
       "      <td>3.095350</td>\n",
       "      <td>6.722305</td>\n",
       "      <td>9.813612</td>\n",
       "      <td>0.0</td>\n",
       "      <td>7.405509</td>\n",
       "      <td>0.499882</td>\n",
       "      <td>0.016744</td>\n",
       "      <td>0.013428</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>std</th>\n",
       "      <td>0.136850</td>\n",
       "      <td>1.200828</td>\n",
       "      <td>1.065601</td>\n",
       "      <td>0.638819</td>\n",
       "      <td>0.506537</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.108237</td>\n",
       "      <td>0.508799</td>\n",
       "      <td>0.133635</td>\n",
       "      <td>0.204722</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>min</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>5.009284</td>\n",
       "      <td>8.435999</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.930747</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25%</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>2.299039</td>\n",
       "      <td>2.390365</td>\n",
       "      <td>6.303346</td>\n",
       "      <td>9.464466</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.676042</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50%</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>3.143687</td>\n",
       "      <td>3.127006</td>\n",
       "      <td>6.655893</td>\n",
       "      <td>9.791599</td>\n",
       "      <td>0.0</td>\n",
       "      <td>7.450114</td>\n",
       "      <td>0.443076</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75%</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>3.883484</td>\n",
       "      <td>3.802534</td>\n",
       "      <td>7.038447</td>\n",
       "      <td>10.142324</td>\n",
       "      <td>0.0</td>\n",
       "      <td>8.121984</td>\n",
       "      <td>0.789354</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>max</th>\n",
       "      <td>1.482332</td>\n",
       "      <td>6.237034</td>\n",
       "      <td>6.063484</td>\n",
       "      <td>10.129528</td>\n",
       "      <td>11.355621</td>\n",
       "      <td>0.0</td>\n",
       "      <td>10.718190</td>\n",
       "      <td>2.779008</td>\n",
       "      <td>1.785592</td>\n",
       "      <td>4.067604</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           gene_0      gene_1      gene_2      gene_3      gene_4  gene_5  \\\n",
       "count  801.000000  801.000000  801.000000  801.000000  801.000000   801.0   \n",
       "mean     0.026642    3.010909    3.095350    6.722305    9.813612     0.0   \n",
       "std      0.136850    1.200828    1.065601    0.638819    0.506537     0.0   \n",
       "min      0.000000    0.000000    0.000000    5.009284    8.435999     0.0   \n",
       "25%      0.000000    2.299039    2.390365    6.303346    9.464466     0.0   \n",
       "50%      0.000000    3.143687    3.127006    6.655893    9.791599     0.0   \n",
       "75%      0.000000    3.883484    3.802534    7.038447   10.142324     0.0   \n",
       "max      1.482332    6.237034    6.063484   10.129528   11.355621     0.0   \n",
       "\n",
       "           gene_6      gene_7      gene_8      gene_9  \n",
       "count  801.000000  801.000000  801.000000  801.000000  \n",
       "mean     7.405509    0.499882    0.016744    0.013428  \n",
       "std      1.108237    0.508799    0.133635    0.204722  \n",
       "min      3.930747    0.000000    0.000000    0.000000  \n",
       "25%      6.676042    0.000000    0.000000    0.000000  \n",
       "50%      7.450114    0.443076    0.000000    0.000000  \n",
       "75%      8.121984    0.789354    0.000000    0.000000  \n",
       "max     10.718190    2.779008    1.785592    4.067604  "
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged_df.iloc[:,0:10].describe()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "79e27f00",
   "metadata": {},
   "source": [
    "# Data Normalization\n",
    "Data normalization is done so that the values are in the same range. This will improve model performance and avoid bias"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "96b89d82",
   "metadata": {},
   "outputs": [],
   "source": [
    "from sklearn.preprocessing import MinMaxScaler, LabelEncoder\n",
    "\n",
    "# Assuming X_train is your feature matrix\n",
    "# Identify categorical columns\n",
    "categorical_cols = [col for col in X_train.columns if X_train[col].dtype == 'object']\n",
    "\n",
    "# Apply label encoding to categorical columns\n",
    "label_encoders = {}\n",
    "for col in categorical_cols:\n",
    "    label_encoders[col] = LabelEncoder()\n",
    "    X_train[col] = label_encoders[col].fit_transform(X_train[col])\n",
    "\n",
    "# Apply MinMax scaling\n",
    "min_max_scaler = MinMaxScaler()\n",
    "X_train_scaled = min_max_scaler.fit_transform(X_train)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d027f9dd",
   "metadata": {},
   "source": [
    "# Feature Selection\n",
    "The purpose of feature selection is to select relevant features for classification. Feature selection is usually used as a pre-processing step before doing the actual learning.\n",
    "\n",
    "In this tutorial, mutual information algorithm is used to compute the relevance of each feature. The top n (eg. 300) features are selected for the machine learning analysis."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ffa70f8d",
   "metadata": {},
   "source": [
    "Mutual information Feature selection"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "38f8f42b",
   "metadata": {},
   "outputs": [],
   "source": [
    "MI=mutual_info_classif(X_train_scaled,y_train)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e8aae9da",
   "metadata": {},
   "source": [
    "Select the top arbitrary n nuber of features "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "d8b798a2",
   "metadata": {},
   "outputs": [],
   "source": [
    "n_features=300\n",
    "selected_scores_indices=np.argsort(MI)[::-1][0:n_features]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "d88b9d9f",
   "metadata": {},
   "outputs": [],
   "source": [
    "X_train_selected=X_train_scaled[:,selected_scores_indices]\n",
    "X_test_scaled = min_max_scaler.transform(X_test)  # Assuming you want to scale X_test\n",
    "X_test_selected = X_test_scaled[:, selected_scores_indices]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "id": "2f23b113",
   "metadata": {},
   "outputs": [
    {
     "ename": "ValueError",
     "evalue": "The feature names should match those that were passed during fit.\nFeature names seen at fit time, yet now missing:\n- Unnamed: 0\n",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "Cell \u001b[1;32mIn[74], line 5\u001b[0m\n\u001b[0;32m      2\u001b[0m X_test_numeric \u001b[38;5;241m=\u001b[39m X_test\u001b[38;5;241m.\u001b[39mdrop(columns\u001b[38;5;241m=\u001b[39mcategorical_cols)\n\u001b[0;32m      4\u001b[0m \u001b[38;5;66;03m# Scale the numeric columns in X_test\u001b[39;00m\n\u001b[1;32m----> 5\u001b[0m X_test_scaled \u001b[38;5;241m=\u001b[39m min_max_scaler\u001b[38;5;241m.\u001b[39mtransform(X_test_numeric)\n\u001b[0;32m      7\u001b[0m \u001b[38;5;66;03m# Select the relevant features\u001b[39;00m\n\u001b[0;32m      8\u001b[0m X_test_selected \u001b[38;5;241m=\u001b[39m X_test_scaled[:, selected_scores_indices]\n",
      "File \u001b[1;32m~\\AppData\\Local\\anaconda3\\Lib\\site-packages\\sklearn\\utils\\_set_output.py:140\u001b[0m, in \u001b[0;36m_wrap_method_output.<locals>.wrapped\u001b[1;34m(self, X, *args, **kwargs)\u001b[0m\n\u001b[0;32m    138\u001b[0m \u001b[38;5;129m@wraps\u001b[39m(f)\n\u001b[0;32m    139\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mwrapped\u001b[39m(\u001b[38;5;28mself\u001b[39m, X, \u001b[38;5;241m*\u001b[39margs, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs):\n\u001b[1;32m--> 140\u001b[0m     data_to_wrap \u001b[38;5;241m=\u001b[39m f(\u001b[38;5;28mself\u001b[39m, X, \u001b[38;5;241m*\u001b[39margs, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs)\n\u001b[0;32m    141\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(data_to_wrap, \u001b[38;5;28mtuple\u001b[39m):\n\u001b[0;32m    142\u001b[0m         \u001b[38;5;66;03m# only wrap the first output for cross decomposition\u001b[39;00m\n\u001b[0;32m    143\u001b[0m         return_tuple \u001b[38;5;241m=\u001b[39m (\n\u001b[0;32m    144\u001b[0m             _wrap_data_with_container(method, data_to_wrap[\u001b[38;5;241m0\u001b[39m], X, \u001b[38;5;28mself\u001b[39m),\n\u001b[0;32m    145\u001b[0m             \u001b[38;5;241m*\u001b[39mdata_to_wrap[\u001b[38;5;241m1\u001b[39m:],\n\u001b[0;32m    146\u001b[0m         )\n",
      "File \u001b[1;32m~\\AppData\\Local\\anaconda3\\Lib\\site-packages\\sklearn\\preprocessing\\_data.py:514\u001b[0m, in \u001b[0;36mMinMaxScaler.transform\u001b[1;34m(self, X)\u001b[0m\n\u001b[0;32m    500\u001b[0m \u001b[38;5;250m\u001b[39m\u001b[38;5;124;03m\"\"\"Scale features of X according to feature_range.\u001b[39;00m\n\u001b[0;32m    501\u001b[0m \n\u001b[0;32m    502\u001b[0m \u001b[38;5;124;03mParameters\u001b[39;00m\n\u001b[1;32m   (...)\u001b[0m\n\u001b[0;32m    510\u001b[0m \u001b[38;5;124;03m    Transformed data.\u001b[39;00m\n\u001b[0;32m    511\u001b[0m \u001b[38;5;124;03m\"\"\"\u001b[39;00m\n\u001b[0;32m    512\u001b[0m check_is_fitted(\u001b[38;5;28mself\u001b[39m)\n\u001b[1;32m--> 514\u001b[0m X \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_validate_data(\n\u001b[0;32m    515\u001b[0m     X,\n\u001b[0;32m    516\u001b[0m     copy\u001b[38;5;241m=\u001b[39m\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mcopy,\n\u001b[0;32m    517\u001b[0m     dtype\u001b[38;5;241m=\u001b[39mFLOAT_DTYPES,\n\u001b[0;32m    518\u001b[0m     force_all_finite\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mallow-nan\u001b[39m\u001b[38;5;124m\"\u001b[39m,\n\u001b[0;32m    519\u001b[0m     reset\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mFalse\u001b[39;00m,\n\u001b[0;32m    520\u001b[0m )\n\u001b[0;32m    522\u001b[0m X \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mscale_\n\u001b[0;32m    523\u001b[0m X \u001b[38;5;241m+\u001b[39m\u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mmin_\n",
      "File \u001b[1;32m~\\AppData\\Local\\anaconda3\\Lib\\site-packages\\sklearn\\base.py:579\u001b[0m, in \u001b[0;36mBaseEstimator._validate_data\u001b[1;34m(self, X, y, reset, validate_separately, cast_to_ndarray, **check_params)\u001b[0m\n\u001b[0;32m    508\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21m_validate_data\u001b[39m(\n\u001b[0;32m    509\u001b[0m     \u001b[38;5;28mself\u001b[39m,\n\u001b[0;32m    510\u001b[0m     X\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mno_validation\u001b[39m\u001b[38;5;124m\"\u001b[39m,\n\u001b[1;32m   (...)\u001b[0m\n\u001b[0;32m    515\u001b[0m     \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mcheck_params,\n\u001b[0;32m    516\u001b[0m ):\n\u001b[0;32m    517\u001b[0m \u001b[38;5;250m    \u001b[39m\u001b[38;5;124;03m\"\"\"Validate input data and set or check the `n_features_in_` attribute.\u001b[39;00m\n\u001b[0;32m    518\u001b[0m \n\u001b[0;32m    519\u001b[0m \u001b[38;5;124;03m    Parameters\u001b[39;00m\n\u001b[1;32m   (...)\u001b[0m\n\u001b[0;32m    577\u001b[0m \u001b[38;5;124;03m        validated.\u001b[39;00m\n\u001b[0;32m    578\u001b[0m \u001b[38;5;124;03m    \"\"\"\u001b[39;00m\n\u001b[1;32m--> 579\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_check_feature_names(X, reset\u001b[38;5;241m=\u001b[39mreset)\n\u001b[0;32m    581\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m y \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m \u001b[38;5;129;01mand\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_get_tags()[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mrequires_y\u001b[39m\u001b[38;5;124m\"\u001b[39m]:\n\u001b[0;32m    582\u001b[0m         \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mValueError\u001b[39;00m(\n\u001b[0;32m    583\u001b[0m             \u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mThis \u001b[39m\u001b[38;5;132;01m{\u001b[39;00m\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m\u001b[38;5;18m__class__\u001b[39m\u001b[38;5;241m.\u001b[39m\u001b[38;5;18m__name__\u001b[39m\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m estimator \u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[0;32m    584\u001b[0m             \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mrequires y to be passed, but the target y is None.\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[0;32m    585\u001b[0m         )\n",
      "File \u001b[1;32m~\\AppData\\Local\\anaconda3\\Lib\\site-packages\\sklearn\\base.py:506\u001b[0m, in \u001b[0;36mBaseEstimator._check_feature_names\u001b[1;34m(self, X, reset)\u001b[0m\n\u001b[0;32m    501\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m missing_names \u001b[38;5;129;01mand\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m unexpected_names:\n\u001b[0;32m    502\u001b[0m     message \u001b[38;5;241m+\u001b[39m\u001b[38;5;241m=\u001b[39m (\n\u001b[0;32m    503\u001b[0m         \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mFeature names must be in the same order as they were in fit.\u001b[39m\u001b[38;5;130;01m\\n\u001b[39;00m\u001b[38;5;124m\"\u001b[39m\n\u001b[0;32m    504\u001b[0m     )\n\u001b[1;32m--> 506\u001b[0m \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mValueError\u001b[39;00m(message)\n",
      "\u001b[1;31mValueError\u001b[0m: The feature names should match those that were passed during fit.\nFeature names seen at fit time, yet now missing:\n- Unnamed: 0\n"
     ]
    }
   ],
   "source": [
    "# Drop non-numeric columns from X_test\n",
    "#X_test_numeric = X_test.drop(columns=categorical_cols)\n",
    "\n",
    "# Scale the numeric columns in X_test\n",
    "#X_test_scaled = min_max_scaler.transform(X_test_numeric)\n",
    "\n",
    "# Select the relevant features\n",
    "#X_test_selected = X_test_scaled[:, selected_scores_indices]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "18ed13aa",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(640, 300)"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "X_train_selected.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "855468c5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(161, 300)"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "X_test_selected.shape\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3e92626f",
   "metadata": {},
   "source": [
    "# Classification"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "aebeb420",
   "metadata": {},
   "source": [
    "Random forest is suitable for handling multiclass and high-dimensional data. The classification process entails both training and testing the model."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "75ec7ff1",
   "metadata": {},
   "source": [
    "# Model Training"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "93fc9b03",
   "metadata": {},
   "outputs": [],
   "source": [
    "RF=OneVsRestClassifier(RandomForestClassifier(max_features=0.2))\n",
    "RF.fit(X_train_selected,y_train)\n",
    "y_pred =RF.predict(X_test_selected)\n",
    "pred_prob = RF.predict_proba(X_test_selected)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "d121619a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>#sk-container-id-1 {color: black;}#sk-container-id-1 pre{padding: 0;}#sk-container-id-1 div.sk-toggleable {background-color: white;}#sk-container-id-1 label.sk-toggleable__label {cursor: pointer;display: block;width: 100%;margin-bottom: 0;padding: 0.3em;box-sizing: border-box;text-align: center;}#sk-container-id-1 label.sk-toggleable__label-arrow:before {content: \"â–¸\";float: left;margin-right: 0.25em;color: #696969;}#sk-container-id-1 label.sk-toggleable__label-arrow:hover:before {color: black;}#sk-container-id-1 div.sk-estimator:hover label.sk-toggleable__label-arrow:before {color: black;}#sk-container-id-1 div.sk-toggleable__content {max-height: 0;max-width: 0;overflow: hidden;text-align: left;background-color: #f0f8ff;}#sk-container-id-1 div.sk-toggleable__content pre {margin: 0.2em;color: black;border-radius: 0.25em;background-color: #f0f8ff;}#sk-container-id-1 input.sk-toggleable__control:checked~div.sk-toggleable__content {max-height: 200px;max-width: 100%;overflow: auto;}#sk-container-id-1 input.sk-toggleable__control:checked~label.sk-toggleable__label-arrow:before {content: \"â–¾\";}#sk-container-id-1 div.sk-estimator input.sk-toggleable__control:checked~label.sk-toggleable__label {background-color: #d4ebff;}#sk-container-id-1 div.sk-label input.sk-toggleable__control:checked~label.sk-toggleable__label {background-color: #d4ebff;}#sk-container-id-1 input.sk-hidden--visually {border: 0;clip: rect(1px 1px 1px 1px);clip: rect(1px, 1px, 1px, 1px);height: 1px;margin: -1px;overflow: hidden;padding: 0;position: absolute;width: 1px;}#sk-container-id-1 div.sk-estimator {font-family: monospace;background-color: #f0f8ff;border: 1px dotted black;border-radius: 0.25em;box-sizing: border-box;margin-bottom: 0.5em;}#sk-container-id-1 div.sk-estimator:hover {background-color: #d4ebff;}#sk-container-id-1 div.sk-parallel-item::after {content: \"\";width: 100%;border-bottom: 1px solid gray;flex-grow: 1;}#sk-container-id-1 div.sk-label:hover label.sk-toggleable__label {background-color: #d4ebff;}#sk-container-id-1 div.sk-serial::before {content: \"\";position: absolute;border-left: 1px solid gray;box-sizing: border-box;top: 0;bottom: 0;left: 50%;z-index: 0;}#sk-container-id-1 div.sk-serial {display: flex;flex-direction: column;align-items: center;background-color: white;padding-right: 0.2em;padding-left: 0.2em;position: relative;}#sk-container-id-1 div.sk-item {position: relative;z-index: 1;}#sk-container-id-1 div.sk-parallel {display: flex;align-items: stretch;justify-content: center;background-color: white;position: relative;}#sk-container-id-1 div.sk-item::before, #sk-container-id-1 div.sk-parallel-item::before {content: \"\";position: absolute;border-left: 1px solid gray;box-sizing: border-box;top: 0;bottom: 0;left: 50%;z-index: -1;}#sk-container-id-1 div.sk-parallel-item {display: flex;flex-direction: column;z-index: 1;position: relative;background-color: white;}#sk-container-id-1 div.sk-parallel-item:first-child::after {align-self: flex-end;width: 50%;}#sk-container-id-1 div.sk-parallel-item:last-child::after {align-self: flex-start;width: 50%;}#sk-container-id-1 div.sk-parallel-item:only-child::after {width: 0;}#sk-container-id-1 div.sk-dashed-wrapped {border: 1px dashed gray;margin: 0 0.4em 0.5em 0.4em;box-sizing: border-box;padding-bottom: 0.4em;background-color: white;}#sk-container-id-1 div.sk-label label {font-family: monospace;font-weight: bold;display: inline-block;line-height: 1.2em;}#sk-container-id-1 div.sk-label-container {text-align: center;}#sk-container-id-1 div.sk-container {/* jupyter's `normalize.less` sets `[hidden] { display: none; }` but bootstrap.min.css set `[hidden] { display: none !important; }` so we also need the `!important` here to be able to override the default hidden behavior on the sphinx rendered scikit-learn.org. See: https://github.com/scikit-learn/scikit-learn/issues/21755 */display: inline-block !important;position: relative;}#sk-container-id-1 div.sk-text-repr-fallback {display: none;}</style><div id=\"sk-container-id-1\" class=\"sk-top-container\"><div class=\"sk-text-repr-fallback\"><pre>OneVsRestClassifier(estimator=RandomForestClassifier(max_features=0.2))</pre><b>In a Jupyter environment, please rerun this cell to show the HTML representation or trust the notebook. <br />On GitHub, the HTML representation is unable to render, please try loading this page with nbviewer.org.</b></div><div class=\"sk-container\" hidden><div class=\"sk-item sk-dashed-wrapped\"><div class=\"sk-label-container\"><div class=\"sk-label sk-toggleable\"><input class=\"sk-toggleable__control sk-hidden--visually\" id=\"sk-estimator-id-1\" type=\"checkbox\" ><label for=\"sk-estimator-id-1\" class=\"sk-toggleable__label sk-toggleable__label-arrow\">OneVsRestClassifier</label><div class=\"sk-toggleable__content\"><pre>OneVsRestClassifier(estimator=RandomForestClassifier(max_features=0.2))</pre></div></div></div><div class=\"sk-parallel\"><div class=\"sk-parallel-item\"><div class=\"sk-item\"><div class=\"sk-label-container\"><div class=\"sk-label sk-toggleable\"><input class=\"sk-toggleable__control sk-hidden--visually\" id=\"sk-estimator-id-2\" type=\"checkbox\" ><label for=\"sk-estimator-id-2\" class=\"sk-toggleable__label sk-toggleable__label-arrow\">estimator: RandomForestClassifier</label><div class=\"sk-toggleable__content\"><pre>RandomForestClassifier(max_features=0.2)</pre></div></div></div><div class=\"sk-serial\"><div class=\"sk-item\"><div class=\"sk-estimator sk-toggleable\"><input class=\"sk-toggleable__control sk-hidden--visually\" id=\"sk-estimator-id-3\" type=\"checkbox\" ><label for=\"sk-estimator-id-3\" class=\"sk-toggleable__label sk-toggleable__label-arrow\">RandomForestClassifier</label><div class=\"sk-toggleable__content\"><pre>RandomForestClassifier(max_features=0.2)</pre></div></div></div></div></div></div></div></div></div></div>"
      ],
      "text/plain": [
       "OneVsRestClassifier(estimator=RandomForestClassifier(max_features=0.2))"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "RF"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "58986614",
   "metadata": {},
   "source": [
    "# Model Evaluation\n",
    "After training the model, it is important to know how the model performs on an independent data(i.e. the test data). The following metrics are used to evaluate the model\n",
    "\n",
    "* accuracy\n",
    "* precision\n",
    "* recall\n",
    "* f1 score\n",
    "* confusion matrix\n",
    "* ROC Curve\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "45ca2579",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "accuracy:0.9875\n",
      "precision:0.9880\n",
      "recall:0.9876\n",
      "f1score:0.9875\n",
      "\n",
      "\n",
      "classification report\n",
      "\n",
      "\n",
      "              precision    recall  f1-score   support\n",
      "\n",
      "        BRCA       0.96      1.00      0.98        55\n",
      "        COAD       1.00      1.00      1.00        17\n",
      "        KIRC       1.00      1.00      1.00        25\n",
      "        LUAD       1.00      0.94      0.97        32\n",
      "        PRAD       1.00      1.00      1.00        32\n",
      "\n",
      "    accuracy                           0.99       161\n",
      "   macro avg       0.99      0.99      0.99       161\n",
      "weighted avg       0.99      0.99      0.99       161\n",
      "\n"
     ]
    }
   ],
   "source": [
    "#accuracy\n",
    "accuracy=np.round(balanced_accuracy_score(y_test,y_pred),4)\n",
    "print('accuracy:%0.4f'%accuracy)\n",
    "\n",
    "#precision\n",
    "precision=np.round(precision_score(y_test,y_pred,average = 'weighted'),4)\n",
    "print('precision:%0.4f'%precision)\n",
    "\n",
    "#recall\n",
    "recall=np.round(recall_score(y_test,y_pred,average = 'weighted'),4)\n",
    "print('recall:%0.4f'%recall)\n",
    "\n",
    "#f1score\n",
    "f1score=np.round(f1_score(y_test,y_pred,average = 'weighted'),4)\n",
    "print('f1score:%0.4f'%f1score)\n",
    "\n",
    "\n",
    "report=classification_report(y_test,y_pred, target_names=labels)\n",
    "print('\\n')\n",
    "print('classification report\\n\\n')\n",
    "print(report)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a4f045b1",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
